src/GraphQL/Internal/Schema.hs

```diff
@@ -32,21 +32,47 @@ module GraphQL.Internal.Schema
   , AnnotatedType(..)
   , ListType(..)
   , NonNullType(..)
+  , DefinesTypes(..)
+  , doesFragmentTypeApply
   ) where
 
 import Protolude hiding (Type)
 
+import qualified Data.Map as Map
 import GraphQL.Value (Value)
 import GraphQL.Internal.Name (HasName(..), Name, unsafeMakeName)
 
 -- XXX: Use the built-in NonEmptyList in Haskell
-newtype NonEmptyList a = NonEmptyList [a] deriving (Eq, Show)
+newtype NonEmptyList a = NonEmptyList [a] deriving (Eq, Show, Functor, Foldable)
+
+-- | A thing that defines types. Excludes definitions of input types.
+class DefinesTypes t where
```

Comment:
`BuildFieldResolver` and `UnionTypeObjectTypeDefinitionList` really stand out in that list.